cowbird/database/stores.py

```diff
@@ -0,0 +1,117 @@
+"""
+Stores to read/write data to from/to `MongoDB` using pymongo.
+"""
+
+import abc
+import logging
+from typing import TYPE_CHECKING
+
+import pymongo
+
+from cowbird.monitoring.monitor import Monitor, MonitorException
+
+if TYPE_CHECKING:
+    from typing import Any, Dict, List, Optional, Tuple
+
+    from pymongo.collection import Collection
+
+LOGGER = logging.getLogger(__name__)
+
+
+class StoreInterface(object, metaclass=abc.ABCMeta):
+    # Store type being used as collection name in mongo and to retrieve a store
+    type = None
+
+    # Fields name used as index inside the mongo collection, with a length > 1, a compound index is created
+    index_fields = []
+
+    def __init__(self):
+        if not self.type:
+            raise NotImplementedError("Store 'type' must be overridden in inheriting class.")
+        if not self.index_fields:
+            raise NotImplementedError("Store 'index_fields' must be overridden in inheriting class.")
+
+
+class MongodbStore:
+    """
+    Base class extended by all concrete store implementations.
+    """
+
+    def __init__(self, collection):
+        # type: (Collection, Optional[Dict[str, Any]]) -> None
+        if not isinstance(collection, pymongo.collection.Collection):
+            raise TypeError("Collection not of expected type.")
+        self.collection = collection  # type: Collection
+
+    @classmethod
+    def get_args_kwargs(cls, *args, **kwargs):
+        # type: (*Any, **Any) -> Tuple[Tuple, Dict]
+        """
+        Filters :class:`MongodbStore`-specific arguments to safely pass them down its ``__init__``.
+        """
+        collection = None
+        if len(args):
+            collection = args[0]
+        elif "collection" in kwargs:    # pylint: disable=R1715
+            collection = kwargs["collection"]
+        return tuple([collection]), {}
+
+
+class MonitoringStore(StoreInterface, MongodbStore):
```

Comment:
I'm not sure inheritance of MongodbStore is the best approach? I'm generally wary of inheritance, especially early in a project (and I don't mean using abstract classes, I completely agree on those). 

Could this need be satisfied with a composition approach or some other mechanism?

Reply:
It may be overkill and at first I avoided it, but I finally tried to stay close to the Weaver approach : https://github.com/crim-ca/weaver/tree/master/weaver/store. When we'll add more stores, having a self.collection prepared by `MongodbStore` will make more sense. Using composition could do the job, but in Python there is not a huge tradeoff between the two.

Reply:
I trust you on this one, I was just taught to avoid inheritance when possible as it has a tendency to complicate things later in the lifecycle.